src/emqx_protocol.erl

```diff
@@ -566,6 +677,22 @@ send(Packet = ?PACKET(Type), PState = #pstate{proto_ver = Ver, sendfun = SendFun
             {error, Reason}
     end.
 
+%% TODO: need a better function name
+open_session(PState = #pstate{keepalive = Keepalive}) ->
+    PState1 = maybe_assign_client_id(PState),
+    case try_open_session(PState1) of
+        {ok, SPid, SP} ->
+            PState2 = PState1#pstate{session = SPid, connected = true},
+            ok = emqx_cm:register_connection(client_id(PState2), attrs(PState2)),
+            %% Start keepalive
+            start_keepalive(Keepalive, PState2),
+            %% Success
+            {?RC_SUCCESS, SP, PState2#pstate{auth_state = connected}};
```

Comment:
if we call `connack` here, it should be a continuation,
not a stack pop for the caller and continue there.

it's also easier to read if continue from here.

then we can convert `connack/1` to `connack/2`.

Reply:
{?RC_SUCCESS, SP, PState} -> connack({?RC_SUCCESS, SP, PState})?